### PR TITLE
Vttablet schema tracking: Fix _vt.schema_version corruption

### PR DESCRIPTION
**Pull request by brendar**
## Description

This PR fixes a race condition which caused protobuf marshaled schema data in `_vt.schema_version` rows to become corrupted when the `ColumnType` of `query.Field` pointers was modified between the time when `Field` message sizes were calculated and when `Field` message data was written to the buffer. 

On Vitess versions <= 16, the race condition leads to invalid protobuf data being written to the `schema_version` table. When the schema historian tries to unmarshal the data it encounters an error, which breaks schema tracking on running tablets, and prevents newly started tablets from serving.

On Vitess main the race condition leads to a panic and the `schema_version` row is never written. This is due to the switch to using `MarshalVT` in https://github.com/vitessio/vitess/pull/12525

I'm proposing to fix this by copying the fields returned from schema.Engine before modifying them. As far as I can tell, nothing relied on the side effect of setting `ColumnType` on the shared fields.

This PR also includes a defensive change to acquire the schema.Engine mutex while marshaling the schema to protobuf. This is not strictly necessary to fix the bug, but it could help avoid future race conditions or ones that haven't been discovered yet. I would be happy to remove it if anyone feels that it's unnecessary.

This bug was pretty disruptive for us, so I think the fix should be backported, but note that if it is backported, it may be desirable to change the call to `MarshalVT` back to `proto.Marshal` to match the existing code in released versions.

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/12981

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/schema/engine.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/engine.go:676]: We know the capacity we need at this point so we should do a single allocation for the needed capacity:
```suggestion
	dbSchema := &binlogdatapb.MinimalSchema{
		Tables: make([]*binlogdatapb.MinimalTable, 0, len(se.tables)),
	}
```

File: go/vt/vttablet/tabletserver/schema/engine.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/engine.go:691]: We should be more efficient here and do a single allocation since we know the size (I know this isn't new, but since we're here...):
```suggestion
	pkc := make([]int64, len(st.PKColumns))
	for i, pk := range st.PKColumns {
		pkc[i] = int64(pk)
	}
```

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:30]: It's very nitty, I know, but can you please arrange it so that the imports are in the following groups and order:
<builtin packages>

<external/3rd party packages>

<vitess.io packages>

**Review by mattlord — COMMENTED** [review]
Looks good! I only had a few minor comments/requests. Please let me know if I'm missing something. I'll quickly approve once that's resolved.

Then we'll need someone from Query Serving to take a quick look as well. 

Thank you for the contribution, @brendar! ❤️

**Comment by deepthi** [comment]
> This bug was pretty disruptive for us, so I think the fix should be backported, but note that if it is backported, it may be desirable to change the call to MarshalVT back to proto.Marshal to match the existing code in released versions.

Agreed. However, the backport will need to be done manually, so I'm going to remove the label.

**Review by harshit-gangal — COMMENTED** [review]
changes look good.

**Review by rohit-nayak-ps — DISMISSED** [review]
Acquiring the schema engine mutex while marshalling makes sense.
Thanks for fixing this tricky race condition!

File: go/vt/vttablet/tabletserver/schema/engine.go
Comment (austenLacy) [reply go/vt/vttablet/tabletserver/schema/engine.go]: done in https://github.com/vitessio/vitess/pull/13045/commits/ee3c5f74f1e95335868b0294f21fd9af59c90689

File: go/vt/vttablet/tabletserver/schema/engine.go
Comment (austenLacy) [reply go/vt/vttablet/tabletserver/schema/engine.go]: done in https://github.com/vitessio/vitess/pull/13045/commits/ee3c5f74f1e95335868b0294f21fd9af59c90689

**Comment by austenLacy** [comment]
@mattlord can we get another 👀 on this when you have a chance? 🙇

**Comment by mattlord** [comment]
@austenLacy can you please merge in `vitessio/main`? That should resolve the unit_race workflow failures.

The static code check workflow is also failing: https://github.com/vitessio/vitess/actions/runs/5070727297/jobs/9106509343?pr=13045

Otherwise it's good! Once the tests are passing I'll approve and merge. ❤️

**Comment by austenLacy** [comment]
@mattlord fixed the `gofmt` issue and rebased off https://github.com/vitessio/vitess/commit/f5a460ca4a73f5750c2262efc5663d9de04a31ac.

```
$ git rev-parse --abbrev-ref HEAD
fix-vt-schema_version-corruption

$ git log --pretty=oneline f5a460ca4a73f5750c2262efc5663d9de04a31ac^..HEAD
8887af88c771dba36d6d65d0cca16479a0ec01e2 (HEAD -> fix-vt-schema_version-corruption) gofmt
50186f23971401a4692a17caaab7f3d8a9e33bf8 reorganize imports and allocate slices with make
7cdbb99196750a2baf6d9bda545c4d37503be51b Acquire schema engine mutex while marshaling schema
607031f753b1421f313d2e9da074fda9a883b2fc Copy fields from schema.Engine before modifying
f5a460ca4a73f5750c2262efc5663d9de04a31ac (origin/main, origin/HEAD) VReplication: Do not delete sharded target vschema table entries on Cancel (#13146)
```

**Comment by mattlord** [comment]
The changes in this PR are causing this test failure:
```
$ go test -count 1 -timeout 30s -run ^TestVStreamCopyCompleteFlow$ vitess.io/vitess/go/vt/vttablet/tabletserver/vstreamer
```

**Comment by austenLacy** [comment]
@rohit-nayak-ps 

Can you take another look at this when you have a moment? There were some [unit tests failing in `uvstreamer_flakey_test.go`](https://github.com/vitessio/vitess/pull/13045#issuecomment-1563251711) that were definitely caused by my changes since they pass on `main`. I fixed them by modifying the expectations in the tests in https://github.com/vitessio/vitess/pull/13045/commits/feba91a4da3671d752a5bdb055881d269a2afd07, but I'm not familiar enough with the `vstreamer` and `uvstreamer` tests to know if that was the right thing to do or not.

Something that I also noticed was that the expected events in the uvstreamer and vstreamer tests are not the same.

#### [vstreamer_flaky_test.go](https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletserver/vstreamer/vstreamer_flaky_test.go) (doesn't have the `column_type` field in the event)
```
"type:FIELD field_event:{table_name:\"t1\" fields:{name:\"id1\" type:INT32 table:\"t1\" org_table:\"t1\" database:\"vttest\" org_name:\"id1\" column_length:11 charset:63} fields:{name:\"id2\" type:INT32 table:\"t1\" org_table:\"t1\" database:\"vttest\" org_name:\"id2\" column_length:11 charset:63}}",
```

#### [uvstreamer_flaky_test.go](https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletserver/vstreamer/uvstreamer_flaky_test.go#L469) (includes the `column_type` field)
```
"type:FIELD field_event:{table_name:\"t1\" fields:{name:\"id11\" type:INT32 table:\"t1\" org_table:\"t1\" database:\"vttest\" org_name:\"id11\" column_length:11 charset:63 column_type:\"int(11)\"} fields:{name:\"id12\" type:INT32 table:\"t1\" org_table:\"t1\" database:\"vttest\" org_name:\"id12\" column_length:11 charset:63 column_type:\"int(11)\"}}",
```

Is it possible the race condition of updating the `ColumnType` on the original `Field` was being worked around in the `uvstreamer` tests by adding it to the expected events?

**Comment by rohit-nayak-ps** [comment]
> Is it possible the race condition of updating the `ColumnType` on the original `Field` was being worked around in the `uvstreamer` tests by adding it to the expected events?

Hmm, if that were true and if _normally_ the `ColumnType` attribute was not set, the test should be erroring out very often, right? That is not what we see in CI.

Also, are you saying that `ColumnType` should not being set in the vstreamer in  that test?  

```
fields = st.Fields[:len(tm.Types)]
	extColInfos, err := vs.getExtColInfos(tm.Name, tm.Database)
	if err != nil {
		return nil, err
	}
	for _, field := range fields {
		// we want the MySQL column type info so that we can properly handle
		// ambiguous binlog events and other cases where the internal types
		// don't match the MySQL column type. One example being that in binlog
		// events CHAR columns with a binary collation are indistinguishable
		// from BINARY columns.
		if extColInfo, ok := extColInfos[field.Name]; ok {
			field.ColumnType = extColInfo.columnType
		}
	}
	return fields, nil
```

**Comment by austenLacy** [comment]
> Also, are you saying that `ColumnType` should not being set in the vstreamer in that test?

I can't say for sure because I'm not super familiar with the differences between vstreamer and uvstreamer. Do you know why some of the events would have the `ColumnType`, but others don't?

**Comment by rohit-nayak-ps** [comment]
@austenLacy, @brendar,  I debugged this branch and figured out the issue. You were right: there is a (planned) side-effect of the previous code that has caused the test to fail. 

The `rowstreamer` also generates vevents in the _copy phase_. It was using the fields directly from the tablet's schema engine tables collection in its `FieldEvent`. The `vstreamer` used to first populate the `ColumnInfo` attribute of this collection in the `vstreamer`s `buildTablePlan`. Now that we are cloning `Fields` in this PR the `rowstreamer` is no longer seeing the `ColumnInfo` since we no longer update the schema engine's tables. 

So we need to modify `rowstreamer` as well to call `getExtColInfos()` and update that attribute before sending the field event. Some amount of refactor will be needed since `getExtColInfos()` is scoped only to `vstreamer` right now. It might also require other unit tests to be modified. 

I can also work on this, but it will be a week or so before I can get to this since we have a `Vitess v17` release coming up and I also have a couple of urgent issues on my plate at the moment.  Let me know ...

**Comment by austenLacy** [comment]
@rohit-nayak-ps thanks for digging into this. Not a rush to get this done ASAP as we have a temporary workaround for now.

> So we need to modify rowstreamer as well to call getExtColInfos() and update that attribute before sending the field event. Some amount of refactor will be needed since getExtColInfos() is scoped only to vstreamer right now. It might also require other unit tests to be modified.

This isn't the cleanest solution, but it was very simple to duplicate the logic for `getExtColInfo` to rowstreamer.

```diff
diff --git a/go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go b/go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go
index 15c8fa5462..081b25fbc9 100644
--- a/go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go
+++ b/go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go
@@ -22,6 +22,8 @@ import (
        "sync"
        "time"
 
+       "google.golang.org/protobuf/proto"
+
        "vitess.io/vitess/go/mysql"
        "vitess.io/vitess/go/mysql/collations"
        "vitess.io/vitess/go/sqltypes"
@@ -148,9 +150,32 @@ func (rs *rowStreamer) buildPlan() error {
        if err != nil {
                return err
        }
+
+       fields := make([]*querypb.Field, len(st.Fields))
+       for i := 0; i < len(fields); i++ {
+               fields[i] = proto.Clone(st.Fields[i]).(*querypb.Field)
+       }
+
+       // TODO: is it OK to use `rs.vschema.keyspace` in place of the database name for the
+       // call to getExtCols?
+       extColInfos, err := rs.getExtColInfos(st.Name, rs.vschema.keyspace)
+       if err != nil {
+               return err
+       }
+       for _, field := range fields {
+               // we want the MySQL column type info so that we can properly handle
+               // ambiguous binlog events and other cases where the internal types
+               // don't match the MySQL column type. One example being that in binlog
+               // events CHAR columns with a binary collation are indistinguishable
+               // from BINARY columns.
+               if extColInfo, ok := extColInfos[field.Name]; ok {
+                       field.ColumnType = extColInfo.columnType
+               }
+       }
+
        ti := &Table{
                Name:   st.Name,
-               Fields: st.Fields,
+               Fields: fields,
        }
        // The plan we build is identical to the one for vstreamer.
        // This is because the row format of a read is identical
@@ -413,3 +438,25 @@ func (rs *rowStreamer) streamQuery(conn *snapshotConn, send func(*binlogdatapb.V
 
        return nil
 }
+
+func (rs *rowStreamer) getExtColInfos(table, database string) (map[string]*extColInfo, error) {
+       extColInfos := make(map[string]*extColInfo)
+       conn, err := rs.cp.Connect(rs.ctx)
+       if err != nil {
+               return nil, err
+       }
+       defer conn.Close()
+       queryTemplate := "select column_name, column_type from information_schema.columns where table_schema=%s and table_name=%s;"
+       query := fmt.Sprintf(queryTemplate, encodeString(database), encodeString(table))
+       qr, err := conn.ExecuteFetch(query, 10000, false)
+       if err != nil {
+               return nil, err
+       }
+       for _, row := range qr.Rows {
+               extColInfo := &extColInfo{
+                       columnType: row[1].ToString(),
+               }
+               extColInfos[row[0].ToString()] = extColInfo
+       }
+       return extColInfos, nil
+}
```

**Comment by austenLacy** [comment]
@rohit-nayak-ps have you had a chance to look at this again? 🙏

**Comment by rohit-nayak-ps** [comment]
> @rohit-nayak-ps have you had a chance to look at this again? pray

I started https://github.com/vitessio/vitess/pull/13285 to update `column_type` in both `vstreamer` and `rowstreamer`. Now that rowstreamer always sends the column_type (as it should always have!) several rowstreamer tests will need to be fixed to allow for it. I have already updated the vstreamer unit tests, but still need to work on the e2e ones and possibly other unit tests.

Once tests pass, we can either 
1. let #13285 flow and get merged and you rebase this PR _OR_ 
2. you can just cherry-pick the commits in #13285 in this PR (after removing the vstreamer changes in this PR). 

Maybe the latter since you have originally started this work? Either is fine with me.

**Comment by austenLacy** [comment]
@rohit-nayak-ps happy to cherry pick when ready 👍

**Comment by austenLacy** [comment]
@rohit-nayak-ps both commits cherry-picked 🙇

**Comment by rohit-nayak-ps** [comment]
@austenLacy, I found a bug in that PR and had to update tests as well to run both on MySQL57 and MySQL80. I squashed all commits into a single one to make it easier to cherry pick.

Please note that you will also need to undo your previous changes to vstreamer and related tests and the commits you already cherry-picked from that PR, before cherry-picking the single commit in #13285.

**Comment by austenLacy** [comment]
@rohit-nayak-ps cherry-picked in https://github.com/vitessio/vitess/pull/13045/commits/85ee37ff6859ead68d3ab8b87cd9e4d9e4dc2411

File: go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go
Comment (brendar) [inline go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go:156]: Passing `rs.vschema.keyspace` may be problematic if the underlying database name is different from the keyspace name.

Comment (rohit-nayak-ps) [reply go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go:156]: Good catch. Sorry, I had caught it in the vstreamer call but forgot to update the rowstreamer one. Fixed here: https://github.com/vitessio/vitess/pull/13285/commits/700e2dd0835073330e9c26d231939420d10cbed6

Comment (austenLacy) [reply go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go]: done in https://github.com/vitessio/vitess/pull/13045/commits/48fd6ee0b7ebf9e2f9a339bae303df55844f616a

**Comment by austenLacy** [comment]
@rohit-nayak-ps I've updated the PR with your PR's latest commits and all checks are ✅ 🎉

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm
Thank you for the PR!

**Review by mattlord — APPROVED** [review]
❤️

**Pull request merged**
